### PR TITLE
Complete allocation algorithm with main network and subnetworks

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/922.

### CODE REVIEW

**Comment by visr** [comment]
The AllocationModels are created once, with Parameter initialization.

**Comment by SouthEndMusic** [comment]
To do:
- [ ] Add convenience function `get_main_network_connections` analogous to `get_allocation_model`;
- [ ] Subnetworks picking up what is allocated to them from the main network doesn't work yet. In `function assign_allocations`, allocations from the main network to a subnetwork should be written to a new field `Allocated.subnetwork_demands::Dict{Tuple{NodeID, NodeID}, Vector{Float64}}`. Currently the allocated value is taken as the optimized flow over the connecting edge in the main network problem, but this does not work at all since this is only the allocated value for the last priority.
- [ ] Complete tests

File: core/src/allocation.jl [hunk 1]
```diff
@@ -952,17 +1037,22 @@ function adjust_edge_capacities!(
...
+
     for edge_id in edge_ids
         c = capacity[edge_id...]
 
-        # Edges with infinite capacity have no capacity constraints
-        if isinf(c)
+        # These edges have no capacity constraints:
+        # - With infinite capacity
+        # - Being a source from the main network to a subnetwork
+        if isinf(c) || edge_id ∈ main_network_source_edges
```
Comment (visr) [inline core/src/allocation.jl:1055]: Seems like `prevfloat` would not work here.

```
julia> isinf(prevfloat(Inf))
false
```

File: core/src/allocation.jl [hunk 2]
```diff
@@ -984,18 +1074,35 @@ end
...
-    (; problem) = allocation_model
+function allocate!(
+    p::Parameters,
+    allocation_model::AllocationModel,
+    t::Float64;
+    collect_demands::Bool = false,
+)::Nothing
+    (; user, allocation) = p
+    (; problem, allocation_network_id) = allocation_model
+    (; allocation_network_ids, main_network_connections) = allocation
```
Comment (visr) [inline core/src/allocation.jl:1085]:
```suggestion
```

**Review by visr — APPROVED** [review]

File: core/src/allocation.jl (hunk 1 above)
Comment (SouthEndMusic) [reply core/src/allocation.jl:1055]: There's the sparse capacity matrix `AllocationModel.capacity` which is determined at initialization which was already allowed to have inifinities, which is related to but not exactly the same as the capacity values in the capacity constraints in the optimization problem. As it turns out, having infinite capacities in these constraints works fine.

**Pull request merged**